java/arcs/core/data/expression/InferredType.kt

```diff
@@ -0,0 +1,123 @@
+package arcs.core.data.expression
+
+import arcs.core.util.BigInt
+import kotlin.reflect.KClass
+
+/**
+ * Matches [widenAndApply] in Expression.kt
+ *
+ * The types are listed in terms of decreasing width, so that a type at position n can always
+ * be widened to a type at position n - 1.
+ */
+private val widenOrder = listOf<InferredType.Primitive>(
+  InferredType.Primitive.NumberType,
+  InferredType.Primitive.DoubleType,
+  InferredType.Primitive.FloatType,
+  InferredType.Primitive.BigIntType,
+  InferredType.Primitive.LongType,
+  InferredType.Primitive.IntType,
+  InferredType.Primitive.ShortType,
+  InferredType.Primitive.ByteType
+)
+
+/** Convert a [KClass] into an appropriate [InferredType] */
+fun KClass<*>.toType() = InferredType.Primitive.of(this)
+
+/** Widen a list of types to the largest type that encompasses all of them. */
+fun widen(vararg types: InferredType): InferredType.Primitive {
+  types.forEach {
```

Comment:
Hmm... why `forEach` here and `for` down there? 

This reminds me a note Maria told me about from the TS codebase: even though `forEach` is an option, `for` is more readable.

Reply:
You can early-exit from for() you can't return from forEach{}. In general, we've been preferring functional operators as a pattern in Kotlin. Not sure what readability thinks.